### ISSUE TITLE
Replaying captured data streams

### ISSUE THREAD

**Issue by pbonte** [SolidLabResearch/Challenges#83]
## Pitch
As many applications are using data streams, a way to easily replay captured streams is necessary for demoing purposes, scalability testing or data recovery. Replaying allows to mimic the real-time behaviour of data streams, even though the data is historical data. A replayer is a crucial component to showcase how our solutions can process live data streams and how they can handle different data rates.
The [DAHCC dataset](https://dahcc.idlab.ugent.be/) will be used as example to replay the data. 

## Desired solution

The replayer should be able to read a number of files, and stream out the events described in each file. To facility performance and scalability testing, the rates and number of streams should be configurable.

## Acceptance criteria
The replayer should be a library that allows to:
- read multiple files or directory,
- stream out the result, possible using multiple output streams (configurable)
- configure the replay frequency, i.e. how fast the events should be streamed out,
- configure burst in the replay frequency, i.e. a certain times the event rate rises drastically to mimic high load/usage periods
- assign new time stamps to the replayed events(the property should be configurable)
- configure the event shape (to know which triples belong to each event)
- Optionally first map raw data, e.g. CSV, to RDF

## Scenarios

This is part of a larger [challenge](https://github.com/SolidLabResearch/Challenges/issues/16)

**Comment by bjdmeest** [comment]
@s-minoo this feels related to your rmlstreamer benchmark

**Comment by s-minoo** [comment]
Yes! I wrote a [data streamer/replayer](https://github.com/s-minoo/datastreamer-rust) in rust which will consume the historical data, and replay them with workload characteristics: periodic burst, constant rate, etc... 

If you want it to apply to this challenge, take a look at two traits of the datastreamer-rust: 
1) [Publisher](https://github.com/s-minoo/datastreamer-rust/blob/main/src/publisher.rs): Responsible for inducing the data stream characteristics: periodic burst, constant rate, etc... 
2) [Processor](https://github.com/s-minoo/datastreamer-rust/blob/main/src/processor.rs): Responsible for parsing the historical data and appending timestamps to the records

**Comment by svrstich** [comment]
Hi, shall certainly have a deeper look into what you've done to see whether it can aligned with the goals of Challenge 82/83.

**Comment by svrstich** [comment]
Loading from pre-configured path currently supported.

**Comment by svrstich** [comment]
Loading of large datasets now supported.
Time-based sorting of measurements in large datasets now supported.
One-step replay now supported.

**Comment by pheyvaer** [comment]
@svrstich Great! What is still missing?

**Comment by svrstich** [comment]
> 

What's still missing? ;-) Bulk replay, tunable bulk-size replay, tunable speed replay, etc.

**Comment by pheyvaer** [comment]
@svrstich Do you have a complete list? So that we can assess a bit better what still needs to happen?

**Comment by svrstich** [comment]
Not really, but we'll be limiting it for the time being to two/three features.

**Comment by svrstich** [comment]
We'll stick with some example parameters for now: step-wise and everything. Other use-case specific replay option can be added later on.

**Comment by svrstich** [comment]
https://gitlab.ilabt.imec.be/svrstich/ldes-in-solid-semantic-observations-replay

**Comment by pheyvaer** [comment]
@svrstich Is this the solution for the challenge, a pointer, or work in progress?

**Comment by pheyvaer** [comment]
@RubenVerborgh Why did you remove "completion: pending" label? Stijn says that it's ready for review.

**Comment by RubenVerborgh** [comment]
My bad; I misunderstood what was written above!

**Comment by pheyvaer** [comment]
Ok, no problem! I assigned it to you now for review.
